07-13-16-language-plugins.md

```diff
@@ -0,0 +1,97 @@
+
+The process for deploying an arbitrary code bundle, regardless of language, to Tessel can be generalized. There are six functions that need to be exported for any language deployment mechanism:
+
+1. Determine Project Location in the file system
+2. Run any compilation/pre-processing steps (optional)
+3. Bundle the executable code into a tarball
+4. Run any pre-execution steps (optional)
+5. Execute the code/binary
+6. Run any post-execution steps (optional)
+
+These functions are more concisely named `checkConfiguration`, `preBundle`, `tarBundle`, `preRun`, `shell`/`binary`, and `postRun`. Here is an explanation of each of those interfaces.
+
+### 1. checkConfiguration
+`checkConfiguration` should be able to accept the directory that was deployed with `t2 run ...` and return the path to the entrypoint of the project. With Rust, we do this by parsing the `Cargo.toml` and with JavaScript we parse the project's `package.json`. It is expected to return an object with a `basename` and optional `program` property (for the name of the project).
+
+```.js
+// checkConfiguration to confirm project deployment files
+exportables.checkConfiguration = (pushdir, basename, program) => {
+  // Should return the details about the project
+  return {
+    basename,
+    program
+  };
+};
+```
+
+### 2. preBundle
+`preBundle` is an optional step for any processes that need to be run prior to the next step where you actually package the directory up to be sent to the Tessel 2. In JavaScript is integral to the binary compilation stage (where we inject our pre-compiled binaries in place of existing project binary dependencies) but it's less critical for Rust where we just set some optional parameters.
+
+```.js
+// Optional function to process options prior to bundling project
+exportables.preBundle = function(options) {
+  // Should return a Promise which gets resolved when the code is ready to be tarred
+  return Promise.resolve();
+};
+```
+
+### 3. tarBundle
+`tarBundle` is a required step which should resolve a `Promise` with a Buffer that represents the tarballed code or binary. The CLI will automatically untar the bundle after it is deployed on the Tessel 2.
+
+In the case of Rust, this was the biggest addition. In order to compile Rust code for Tessel, we needed to set up a special cross-compiler that could compile a binary that runs on Tessel's MIPS architecture. The easiest solution was to have the CLI send out the project to a preconfigured remote [cross-compilation server](https://github.com/tessel/rust-compilation-server). The server then compiles and returns the binary to the CLI.
+
+```.js
+// This must implement a Promise that resolves with a Buffer
```

Comment:
I'm writing it for anyone that wants to implement support for a new language. That means they'll create a new file, like `scala.js`, and re-implement all these functions that I'm outlining. There is a copy of each of these functions for each language. The author is free to choose how they implement the function but it won't work if they don't return a Promise, hence the "this must".